### DISCUSSION TITLE
Are there any rules for splitting the OCIRepository or GitRepository?

### DISCUSSION THREAD

**Discussion by ksudate** [fluxcd/flux2#4100] [Q&A]
Hi!

In my environment, I create as many GitRepository as the number of Kustomization.
However, I also know that one GitRepository will work since they are all configured the same.

What rules do you follow when you create a GitRepository?
Do you create one for each Kustomization like I do?

Also, when you use OCIRepository, what rules do you use to divide OCI Registry?
For example, if you need to install multiple add-ons on a cluster, do you store them in the OCI Registry?
Or do you split the Registry?

**Comment by makkes** [comment]
If all your Kustomizations live in the same Namespace then creating multiple GitRepositories with the exact same spec will hurt performance or at least waste resources because source-controller would create multiple clones of the same repository and reconcile them. In that case I'd advise to create only one GitRepository.

As for more general advice I'd check out the [example repository here](https://github.com/fluxcd/flux2-kustomize-helm-example).

**Reply by ksudate** [reply]
@makkes 
Thanks for your answer!

In [apps.yaml](https://github.com/fluxcd/flux2-kustomize-helm-example/blob/main/clusters/staging/apps.yaml), it seems that multiple Controllers are managed in one Kustomization. If the number of Controllers increases, will the Sync time be prolonged?
Would Sync be faster if Kustomization is prepared for each Controller?

**Reply by stefanprodan** [reply]
Having one Flux Kustomization per app/controller is best, but if the controllers are deployed with HelmReleases instead of plain yamls, it's Ok to have a single Kusatomization for all controllers because you can suspend/resume/debug/etc each HelmRelease. At minimum you should have 3 Kustomizations like showed here https://github.com/fluxcd/flux2-kustomize-helm-example/tree/main/clusters/production

**Reply by ksudate** [reply]
@stefanprodan 
Thank you for your response!
It was very helpful.

Finally, I would like to ask about OCIRepository.
I understood that it is better to create only one GitRepository for the same Namespace with the same settings.
What about OCIRepository?
Is it better to have one OCI Registry for each application/controller as well as Kustomization?
Or is it better to store manifests of multiple application/controller in one OCI Registry?

**Reply by stefanprodan** [reply]
OCIRepository is interchangeable with GitRepository, you can do `flux push` from the root of your Git repo and all manifests are pushed to a container registry. The main advantage is that with OCI you don't need to setup SSH, if you're using a cloud supported by Flux, then auth is passwordless as Flux can use the same OIDC as kubelet.

**Reply by ksudate** [reply]
@stefanprodan 
I see that the OCIRepository stores all manifests as well as the GitRepository.

In my use case, I have multiple app/controller k8s manifests (plain yamls) in one repository.
Therefore, if I store all manifests in the OCIRepository, I will not be able to release them individually.
Of course, there is a way to suspend/resume... in Kustomization, but it is not practical to perform this operation every time.

Therefore, we create an OCIRepository for each app/controller and save the manifests.

Are there any problems with this method?
Is there another way?

**Reply by stefanprodan** [reply]
> Therefore, we create an OCIRepository for each app/controller and save the manifests.

That’s how I would do it. Having a dedicated OCI artifact for each deployable unit not only makes life easier to ops/on-call people but also gives devs better control over versioning/rollback. There is no downside to this, unless you deploy tens of thousands of apps from a single registry. At massive scale, Flux could be throttled, as most container registries have rate limits in place, to avoid these, you need to configure Flux webhook receivers and bump the OCIReposities reconciliation interval.

**Reply by ksudate** [reply]
Thanks for the quick answer!
I learned a lot.

I wasn't sure how to configure the OCIRepository, so it helped!
